Timestamp: 19:04 14/05/2026

Câu hỏi đặt ra:

1. Phụ nữ có tỉ lệ sống sót cao hơn đàn ông không?
-> Có

2. Hạng vé 1 có tỉ lệ sống sót cao hơn hạng vé 3 không?
-> Có, vì có thể các khoang hạng 1 sẽ cao cấp và an toàn hơn hạng 3

3. Tỉ lệ sống sót toàn tàu là bao nhiêu %?
-> Khoảng 20%

In [28]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sb

df = sb.load_dataset('titanic')
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


** Bảng có các cột survived (0/1), pclass (1/2/3), sex, age, fare...
Cột 'survived' = 1 nghĩa là sống sót, = 0 là không sống sót.
Ô NaN ở cột 'age' → có missing values.

In [13]:
df.tail()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


** 891 hàng, 15 cột. 
Các cột kiểu object: sex, embarked, class, who, embark_town, alive, alone.
Các cột kiểu float64: age, fare → có thể có NaN vì float mới lưu NaN được.

In [14]:
df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


Tuổi (age): từ 0.42 đến 80 tuổi, trung bình 29.7 tuổi
Giá vé (fare): từ 0 đến 512, chênh lệch rất lớn → có thể có outlier
Survived: mean = 0.38 → tỉ lệ sống sót khoảng 38%

In [15]:
# Kiểm tra missing values

missing_count = df.isnull().sum()
missing_percentage = (df.isnull().sum() / len(df) * 100).round(2)

missing_df = pd.DataFrame({
    'Số ô thiếu': missing_count,
    'Tỉ lệ thiếu (%)': missing_percentage
}).sort_values('Tỉ lệ thiếu (%)', ascending=False)
print(missing_df[missing_df['Số ô thiếu'] > 0])

             Số ô thiếu  Tỉ lệ thiếu (%)
deck                688            77.22
age                 177            19.87
embarked              2             0.22
embark_town           2             0.22


- 'deck': thiếu 688/891 = 77.2%
- 'age': thiếu 177/891 = 19.9%

In [16]:
# Phân tích cột categorical
print("sex")
print(df['sex'].value_counts())
print("\npclass")
print(df['pclass'].value_counts())
print("\nembarked")
print(df['embarked'].value_counts())

sex
sex
male      577
female    314
Name: count, dtype: int64

pclass
pclass
3    491
1    216
2    184
Name: count, dtype: int64

embarked
embarked
S    644
C    168
Q     77
Name: count, dtype: int64


- sex: 577 nam (64.7%) vs 314 nữ (35.3%)
- pclass: hạng 3 đông nhất (491 người), hạng 1 ít nhất (216 người)
- embarked: S (Southampton) đông nhất với 644 người

In [18]:
# Phân tích cột numeric
for columns in ['age', 'fare', 'sibsp']:
    print(f"\n{columns}")
    print(f"  Mean:   {df[columns].mean():.2f}")
    print(f"  Median: {df[columns].median():.2f}")
    print(f"  Min:    {df[columns].min():.2f}")
    print(f"  Max:    {df[columns].max():.2f}")


age
  Mean:   29.70
  Median: 28.00
  Min:    0.42
  Max:    80.00

fare
  Mean:   32.20
  Median: 14.45
  Min:    0.00
  Max:    512.33

sibsp
  Mean:   0.52
  Median: 0.00
  Min:    0.00
  Max:    8.00


- fare: max = 512.33, nhưng median chỉ 14.45
- sibsp (anh chị em/vợ chồng): max = 8
- age: mean (29.7) > median (28)

In [26]:
# 1. Phụ nữ có tỉ lệ sống sót cao hơn đàn ông không?
ti_le_gioi_tinh = df.groupby('sex')['survived'].mean()
print("Tỉ lệ sống sót theo giới tính: ")
print(ti_le_gioi_tinh)

Tỉ lệ sống sót theo giới tính: 
sex
female    0.742038
male      0.188908
Name: survived, dtype: float64


- Phụ nữ: ~74% sống sót
- Đàn ông: ~19% sống sót
→ Dự đoán ĐÚNG. Chênh lệch rất lớn — gần 4 lần.
Nguyên nhân: nguyên tắc "women and children first" được áp dụng 
khi phân phối xuồng cứu sinh.

In [20]:
# 2. Hạng vé 1 có tỉ lệ sống sót cao hơn hạng 3 không?
ti_le_hang_ve = df.groupby('pclass')['survived'].mean()
print("Tỉ lệ sống sót theo hạng vé: ")
print(ti_le_hang_ve)

Tỉ lệ sống sót theo hạng vé: 
pclass
1    0.629630
2    0.472826
3    0.242363
Name: survived, dtype: float64


- Hạng 1: ~63% sống sót
- Hạng 3: ~24% sống sót
→ Dự đoán ĐÚNG. Hạng vé ảnh hưởng rõ rệt.
Không chỉ vì cabin an toàn hơn mà còn vì:
khoang hạng 1 gần boong tàu hơn → tiếp cận xuồng cứu sinh dễ hơn.
Hạng 3 ở tầng dưới → khó thoát ra ngoài khi tàu chìm.

In [24]:
# 3. Tỉ lệ sống sót toàn tàu là bao nhiêu %?
tong_nguoi = len(df)
so_nguoi_song_sot = df['survived'].sum()
so_nguoi_chet = tong_nguoi - so_nguoi_song_sot
ti_le = df['survived'].mean()

print("Tỉ lệ sống sót toàn tàu")
print(f"Tổng số hành khách : {tong_nguoi}")
print(f"Số người sống sót  : {so_nguoi_song_sot}")
print(f"Số người tử nạn    : {so_nguoi_chet}")
print(f"Tỉ lệ sống sót     : {ti_le: }")

Tỉ lệ sống sót toàn tàu
Tổng số hành khách : 891
Số người sống sót  : 342
Số người tử nạn    : 549
Tỉ lệ sống sót     :  0.3838383838383838


- Tỉ lệ sống sót thực tế: ~38%
→ Dự đoán SAI. Tôi đoán ~20% nhưng thực tế cao hơn gần gấp đôi.
Lý do có thể sai: tưởng thảm họa khiến hầu hết không thoát được,
nhưng thực ra xuồng cứu sinh đã cứu được một phần đáng kể —
đặc biệt là phụ nữ và trẻ em.

## Đối chiếu dự đoán vs thực tế

| # | Câu hỏi | Dự đoán | Thực tế | Đúng/Sai |
|---|---------|---------|---------|-----------|
| 1 | Phụ nữ sống nhiều hơn đàn ông? | Có | Có (74% vs 19%) | ✅ Đúng |
| 2 | Hạng 1 sống nhiều hơn hạng 3? | Có | Có (63% vs 24%) | ✅ Đúng |
| 3 | Tỉ lệ sống toàn tàu ~20%? | ~20% | 38.4% | ❌ Sai |

**Đúng 2/3 dự đoán. Sai ở câu 3 — 
con số thực cao hơn dự đoán vì đã có ưu tiên rõ ràng 
cho phụ nữ và trẻ em khi phân phối xuồng cứu sinh.

# Câu hỏi mới sau khi phân tích

1. Tỉ lệ sống của phụ nữ hạng 3 so với đàn ông hạng 1 như thế nào?

2. Cabin (deck) có liên quan đến tỉ lệ sống sót không?

3. Port of embarkation (nơi lên tàu) có ảnh hưởng đến tỉ lệ sống sót không?